tests/core/resource/test_record.py

```diff
@@ -227,6 +227,134 @@ def test_patch_record_returns_updated_fields(self):
+            {'op': 'remove', 'path': '/data/b/1'},
+        ]
+        result = self.resource.patch()['data']
+        self.assertEqual(len(result['b']), 1)
+
+        self.resource.request.json = [
+            {'op': 'test', 'path': '/data/b/0', 'value': 'bbb'},
+            {'op': 'remove', 'path': '/data/b/0'},
+        ]
+        self.assertRaises(httpexceptions.HTTPBadRequest, self.resource.patch)
+
+        self.resource.request.json = [
+            {'op': 'test', 'path': '/data/f', 'value': 'what'},
+        ]
+        self.assertRaises(httpexceptions.HTTPBadRequest, self.resource.patch)
+
+    def test_json_patch_move(self):
+        self.resource.request.json = {'data': {'a': 'aaa', 'b': ['bb','bbb'], 'd': []}}
+        self.resource.patch()
+        header = self.resource.request.headers
+        header['Content-Type'] = 'application/json-patch+json'
+        self.resource.request.json = [
+            {'op': 'move', 'from': '/data/a', 'path': '/data/c'},
+            {'op': 'move', 'from': '/data/b/1', 'path': '/data/d/0'},
+            {'op': 'move', 'from': '/data/b/0', 'path': '/data/e'},
+        ]
+        result = self.resource.patch()['data']
+        self.assertNotIn('a', result)
+        self.assertEqual(result['c'], 'aaa')
+        self.assertEqual(len(result['b']), 0)
+        self.assertEqual(result['d'][0], 'bbb')
+        self.assertEqual(result['e'], 'bb')
+
+        self.resource.request.json = [
+            {'op': 'move', 'from': '/data/f', 'path': '/data/what'},
+        ]
+        self.assertRaises(httpexceptions.HTTPBadRequest, self.resource.patch)
+
+    def test_json_patch_copy(self):
+        self.resource.request.json = {'data': {'a': 'aaa', 'b': ['bb','bbb'], 'd': []}}
+        self.resource.patch()
+        header = self.resource.request.headers
+        header['Content-Type'] = 'application/json-patch+json'
+        self.resource.request.json = [
+            {'op': 'copy', 'from': '/data/a', 'path': '/data/c'},
+            {'op': 'copy', 'from': '/data/b/1', 'path': '/data/d/0'},
+            {'op': 'copy', 'from': '/data/d/0', 'path': '/data/e'},
+        ]
+        result = self.resource.patch()['data']
+        self.assertEqual(result['a'], 'aaa')
+        self.assertEqual(result['c'], 'aaa')
+        self.assertEqual(len(result['b']), 2)
+        self.assertEqual(result['d'][0], 'bbb')
+        self.assertEqual(result['e'], 'bbb')
+
+        self.resource.request.json = [
+            {'op': 'copy', 'from': '/data/f', 'path': '/data/what'},
+        ]
+        self.assertRaises(httpexceptions.HTTPBadRequest, self.resource.patch)
+
+    def test_json_patch_format_not_accepted_without_header(self):
+        self.resource.request.json = {'data': {'a': 'aaa'}}
+        self.resource.patch()
+        self.resource.request.json = [
+            {'op': 'copy', 'from': '/data/a', 'path': '/data/c'},
+        ]
+        self.assertRaises(AttributeError, self.resource.patch)
```

Comment:
This means the client will receive a `500 Internal Error` instead of a clean `400` error response